### PR TITLE
API Strongly type methods

### PR DESCRIPTION
**Pull request by emteknetnz**
**DO NOT MERGE**

Issue https://github.com/silverstripe/silverstripe-framework/issues/10414

This is only a POC

It's automatic code-writing using https://github.com/emteknetnz/type-transitioner which is a 3-step process:

1. Code-writing to add trace statements `_a()` and `_r()` to all methods which are used for log trace results to a text file
2. Running ALL unit-test and behat tests in recipe-kitchen-sink to generate the trace text file
3. Using the results of the trace text file, code write in strongly type params and return types

This was done as a spike, so there's still plenty to do e.g. this code won't run, as you'll soon be greeted with the exception `
Fatal error: Declaration of SilverStripe\Control\HTTPRequest::offsetExists(string $offset) must be compatible with ArrayAccess::offsetExists(mixed $offset): bool in /var/www/vendor/silverstripe/framework/src/Control/HTTPRequest.php on line 438` - that's not a big deal, we'd just need to make the code writer more sophisticated.

### CODE REVIEW

File: src/Control/Controller.php [hunk 1]
```diff
@@ -142,7 +142,7 @@ public function doInit()
      *
      * Also set the URLParams
      */
-    public function setRequest($request)
+    public function setRequest(SilverStripe\Control\NullHTTPRequest $request): SilverStripe\CMS\Controllers\CMSMain
```
Comment (maxime-rainville) [inline src/Control/Controller.php:145]: How could this be the return type here?

File: src/Control/Controller.php [hunk 2]
```diff
@@ -228,7 +228,7 @@ public function handleRequest(HTTPRequest $request)
      *
      * @param HTTPResponse|Object $response
      */
-    protected function prepareResponse($response)
+    protected function prepareResponse(string|SilverStripe\ORM\FieldType\DBHTMLText $response): void
```
Comment (maxime-rainville) [inline src/Control/Controller.php:231]: Type here looks wrong.

File: src/Control/Controller.php [hunk 3]
```diff
@@ -350,7 +350,7 @@ public function getResponse()
      *
      * @return $this
      */
-    public function setResponse(HTTPResponse $response)
+    public function setResponse(HTTPResponse $response): SilverStripe\Dev\DevelopmentAdmin
```
Comment (maxime-rainville) [inline src/Control/Controller.php:353]: Return type looks wrong.

File: src/Control/Controller.php [hunk 4]
```diff
@@ -550,7 +550,7 @@ public function disableBasicAuth()
      *
      * @return Controller
      */
-    public static function curr()
+    public static function curr(): SilverStripe\ErrorPage\ErrorPageController
```
Comment (maxime-rainville) [inline src/Control/Controller.php:553]: This looks wrong.

**Review by maxime-rainville — COMMENTED** [review]
Just looking at `Controller`, there's 4 horrible type mismatches.

Is the tool only looking at what the unit test are using to call the method to try to determine what the expected types are? If that's the case, that means that we'll probably end up replicating whatever errors are in the unit tests or not pick types that generic enough.

**Comment by emteknetnz** [comment]
I left a comment https://github.com/silverstripe/silverstripe-framework/issues/10414#issuecomment-1221669651 - where I later realised there was an issue with the method to get common ancestors that caused an error with a number of arg/return types.  Fixing that bug would mean that something like `Controller` was returned

However, even if that wasn't an issue, will still have poor results like this which are a limitation of taking a trace-only approach:

image

Since we're limited by unit-test coverage, we'd could add a static code analyser to get cases such as this.

Regardless though, this approach has been deemed too risky for CMS 5, so I'm closing this PR

**Pull request closed without merging**